File: api/facts.py
"""
Lightweight scoped fact and memory-candidate store for Hermes WebUI.

Safety rules for Batch 1:
- local JSONL only under ~/.hermes/webui/facts by default;
- no Paperclip writes;
- no Telegram sends;
- no durable Hermes memory writes.
"""
from __future__ import annotations

import json
import os
import threading
import time
import uuid
from pathlib import Path
from typing import Any, Iterable

try:
    from api.config import STATE_DIR
except Exception:  # pragma: no cover - import fallback for isolated tests
    STATE_DIR = Path.home() / '.hermes' / 'webui'

FACTS_DIR = Path(os.getenv('HERMES_WEBUI_FACTS_DIR', str(STATE_DIR / 'facts'))).expanduser().resolve()
CANDIDATES_FILE = 'candidates.jsonl'
FACTS_FILE = 'facts.jsonl'
_LOCK = threading.Lock()

_ALLOWED_CATEGORIES = {'decision', 'preference', 'pattern', 'knowledge', 'constraint'}
_ALLOWED_SCOPES = {'global', 'company', 'project', 'telegram_group', 'workspace', 'profile'}
_ALLOWED_SENSITIVITY = {'public', 'internal', 'confidential'}
_ALLOWED_ACTIONS = {'approve', 'edit', 'reject', 'paperclip_draft_only'}


def _now() -> float:
    return time.time()


def _store_dir(store_dir: str | Path | None = None) -> Path:
    return Path(store_dir).expanduser().resolve() if store_dir else FACTS_DIR


def _path(name: str, store_dir: str | Path | None = None) -> Path:
    base = _store_dir(store_dir)
    base.mkdir(parents=True, exist_ok=True)
    return base / name


def _read_jsonl(path: Path) -> list[dict[str, Any]]:
    if not path.exists():
        return []
    rows: list[dict[str, Any]] = []
    for line in path.read_text(encoding='utf-8').splitlines():
        raw = line.strip()
        if not raw:
            continue
        try:
            obj = json.loads(raw)
        except Exception:
            continue
        if isinstance(obj, dict):
            rows.append(obj)
    return rows


def _write_jsonl(path: Path, rows: Iterable[dict[str, Any]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    text = ''.join(json.dumps(row, ensure_ascii=False, sort_keys=True) + '\n' for row in rows)
    path.write_text(text, encoding='utf-8')


def _append_jsonl(path: Path, row: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open('a', encoding='utf-8') as fh:
        fh.write(json.dumps(row, ensure_ascii=False, sort_keys=True) + '\n')


def _clean_category(value: str | None) -> str:
    value = (value or 'knowledge').strip()
    return value if value in _ALLOWED_CATEGORIES else 'knowledge'


def _clean_scope(value: str | None) -> str:
    value = (value or 'global').strip()
    return value if value in _ALLOWED_SCOPES else 'global'


def create_candidate(
    category: str,
    scope: str,
    scope_ref: str,
    statement: str,
    source_session_id: str | None = None,
    source_message_ids: list[str] | None = None,
    confidence: float | None = None,
    reason: str | None = None,
    sensitivity: str = 'internal',
    recommended_action: str = 'approve',
    metadata: dict[str, Any] | None = None,
    store_dir: str | Path | None = None,
) -> dict[str, Any]:
    statement = str(statement or '').strip()
    if not statement:
        raise ValueError('statement is required')
    if confidence is None:
        confidence = 0.5
    confidence = max(0.0, min(1.0, float(confidence)))
    sensitivity = sensitivity if sensitivity in _ALLOWED_SENSITIVITY else 'internal'
    recommended_action = recommended_action if recommended_action in _ALLOWED_ACTIONS else 'approve'
    now = _now()
    row = {
        'id': 'cand_' + uuid.uuid4().hex[:16],
        'category': _clean_category(category),
        'scope': _clean_scope(scope),
        'scope_ref': str(scope_ref or 'default').strip() or 'default',
        'statement': statement,
        'source_session_id': source_session_id,
        'source_message_ids': source_message_ids or [],
        'confidence': confidence,
        'sensitivity': sensitivity,
        'recommended_action': recommended_action,
        'reason': str(reason or '').strip(),
        'status': 'pending',
        'created_at': now,
        'updated_at': now,
        'metadata': metadata or {},
        'safety': {
            'paperclip_write': False,
            'telegram_send': False,
            'durable_memory_write': False,
        },
    }
    with _LOCK:
        _append_jsonl(_path(CANDIDATES_FILE, store_dir), row)
    return row


def list_candidates(status: str | None = 'pending', scope: str | None = None, scope_ref: str | None = None, store_dir: str | Path | None = None) -> list[dict[str, Any]]:
    rows = _read_jsonl(_path(CANDIDATES_FILE, store_dir))
    if status and status != 'all':
        rows = [r for r in rows if r.get('status') == status]
    if scope:
        rows = [r for r in rows if r.get('scope') == scope]
    if scope_ref:
        rows = [r for r in rows if r.get('scope_ref') == scope_ref]
    return sorted(rows, key=lambda r: r.get('created_at', 0), reverse=True)


def _replace_candidate(candidate_id: str, updater, store_dir: str | Path | None = None) -> dict[str, Any]:
    path = _path(CANDIDATES_FILE, store_dir)
    rows = _read_jsonl(path)
    for idx, row in enumerate(rows):
        if row.get('id') == candidate_id:
            rows[idx] = updater(dict(row))
            _write_jsonl(path, rows)
            return rows[idx]
    raise KeyError('candidate not found')


def approve_candidate(candidate_id: str, edited_statement: str | None = None, store_dir: str | Path | None = None) -> dict[str, Any]:
    with _LOCK:
        def mark(row):
            if row.get('status') != 'pending':
                raise ValueError('candidate is not pending')
            row['status'] = 'approved'
            if edited_statement is not None and str(edited_statement).strip():
                row['statement'] = str(edited_statement).strip()
                row['edited'] = True
            row['updated_at'] = _now()
            return row
        candidate = _replace_candidate(candidate_id, mark, store_dir)
        now = _now()
        fact = {
            'id': 'fact_' + uuid.uuid4().hex[:16],
            'category': candidate.get('category', 'knowledge'),
            'scope': candidate.get('scope', 'global'),
            'scope_ref': candidate.get('scope_ref', 'default'),
            'statement': candidate.get('statement', ''),
            'source_candidate_id': candidate.get('id'),
            'source_session_id': candidate.get('source_session_id'),
            'source_message_ids': candidate.get('source_message_ids', []),
            'confidence': candidate.get('confidence'),
            'sensitivity': candidate.get('sensitivity', 'internal'),
            'status': 'active',
            'relations': [],
            'created_at': now,
            'updated_at': now,
            'metadata': candidate.get('metadata', {}),
            'safety': {
                'paperclip_write': False,
                'telegram_send': False,
                'durable_memory_write': False,
            },
        }
        _append_jsonl(_path(FACTS_FILE, store_dir), fact)
    return {'candidate': candidate, 'fact': fact}


def reject_candidate(candidate_id: str, reason: str | None = None, store_dir: str | Path | None = None) -> dict[str, Any]:
    with _LOCK:
        def mark(row):
            if row.get('status') != 'pending':
                raise ValueError('candidate is not pending')
            row['status'] = 'rejected'
            row['rejection_reason'] = str(reason or '').strip()
            row['updated_at'] = _now()
            return row
        return _replace_candidate(candidate_id, mark, store_dir)


def list_facts(scope: str | None = None, scope_ref: str | None = None, category: str | None = None, query: str | None = None, status: str = 'active', store_dir: str | Path | None = None) -> list[dict[str, Any]]:
    rows = _read_jsonl(_path(FACTS_FILE, store_dir))
    if status and status != 'all':
        rows = [r for r in rows if r.get('status') == status]
    if scope:
        rows = [r for r in rows if r.get('scope') == scope]
    if scope_ref:
        rows = [r for r in rows if r.get('scope_ref') == scope_ref]
    if category:
        rows = [r for r in rows if r.get('category') == category]
    if query:
        q = query.casefold()
        rows = [r for r in rows if q in str(r.get('statement', '')).casefold() or q in str(r.get('scope_ref', '')).casefold()]
    return sorted(rows, key=lambda r: r.get('created_at', 0), reverse=True)


def store_summary(store_dir: str | Path | None = None) -> dict[str, Any]:
    candidates = _read_jsonl(_path(CANDIDATES_FILE, store_dir))
    facts = _read_jsonl(_path(FACTS_FILE, store_dir))
    pending = sum(1 for c in candidates if c.get('status') == 'pending')
    return {
        'store_dir': str(_store_dir(store_dir)),
        'candidates': len(candidates),
        'pending_candidates': pending,
        'facts': len(facts),
        'safety': 'local-only; no Paperclip/Telegram/durable-memory writes',
    }

File: static/facts.js
// Memory Candidate Inbox MVP (Batch 1)
let _memoryInboxCache = {candidates: [], facts: []};

async function loadMemoryInbox(){
  const box=$('memoryInboxPanel');
  if(!box)return;
  box.innerHTML='<div class="memory-inbox-empty">불러오는 중...</div>';
  try{
    const [candData, factData]=await Promise.all([
      api('/api/memory-candidates?status=pending'),
      api('/api/facts?status=active')
    ]);
    _memoryInboxCache.candidates=candData.candidates||[];
    _memoryInboxCache.facts=factData.facts||[];
    renderMemoryInbox(candData.summary||factData.summary||{});
  }catch(e){
    box.innerHTML=`<div class="memory-inbox-error">Memory Inbox 로드 실패: ${esc(e.message||String(e))}</div>`;
  }
}

function renderMemoryInbox(summary){
  const box=$('memoryInboxPanel');
  if(!box)return;
  const candidates=_memoryInboxCache.candidates||[];
  const facts=_memoryInboxCache.facts||[];
  const summaryHtml=`
    <div class="memory-inbox-summary">
      <div><strong>${candidates.length}</strong><span>승인 대기</span></div>
      <div><strong>${facts.length}</strong><span>활성 Fact</span></div>
      <div><strong>승인형</strong><span>Paperclip/Telegram 자동 반영 없음</span></div>
    </div>`;
  if(!candidates.length){
    box.innerHTML=summaryHtml + '<div class="memory-inbox-empty">대기 중인 기억 후보가 없습니다. 오른쪽 빠른 실행의 “기억 후보 추출”로 후보 JSON을 만들고, 하단 입력 폼에 붙여 넣어 저장하세요.</div>' + memoryCandidateFormHtml() + renderFactsPreview(facts);
    return;
  }
  box.innerHTML=summaryHtml + candidates.map(renderMemoryCandidateCard).join('') + memoryCandidateFormHtml() + renderFactsPreview(facts);
}

function renderMemoryCandidateCard(c){
  const badge=c.recommended_action==='paperclip_draft_only' ? '<span class="memory-badge draft">Paperclip draft only</span>' : `<span class="memory-badge">${esc(c.category||'knowledge')}</span>`;
  return `<div class="memory-candidate-card" data-candidate-id="${esc(c.id)}">
    <div class="memory-candidate-head">
      <div>${badge}<span class="memory-badge scope">${esc(c.scope||'global')}:${esc(c.scope_ref||'default')}</span></div>
      <span class="memory-confidence">${Math.round(Number(c.confidence||0)*100)}%</span>
    </div>
    <textarea class="memory-candidate-statement" id="candEdit_${esc(c.id)}">${esc(c.statement||'')}</textarea>
    <div class="memory-candidate-meta">
      source: ${esc(c.source_session_id||'manual')} · ${esc(c.reason||'승인 후에만 fact store에 저장됩니다.')}
    </div>
    <div class="memory-candidate-actions">
      <button class="cron-btn run" onclick="approveMemoryCandidate('${esc(c.id)}')">승인 → Fact</button>
      <button class="cron-btn" onclick="rejectMemoryCandidate('${esc(c.id)}')">거절</button>
    </div>
  </div>`;
}

function memoryCandidateFormHtml(){
  return `<div class="memory-candidate-form">
    <div class="memory-candidate-form-title">수동 후보 추가</div>
    <textarea id="memoryCandidateJson" rows="5" placeholder='JSON 1개 또는 배열을 붙여넣기: {"category":"decision","scope":"project","scope_ref":"...","statement":"..."}'></textarea>
    <div class="memory-candidate-actions">
      <button class="cron-btn run" onclick="submitMemoryCandidateJson()">후보 저장</button>
      <button class="cron-btn" onclick="fillMemoryCandidateExample()">예시</button>
    </div>
    <div class="memory-inbox-note">안전 원칙: 이 기능은 로컬 후보/Facts JSONL만 씁니다. memory tool, Paperclip, Telegram은 실행하지 않습니다.</div>
  </div>`;
}

function renderFactsPreview(facts){
  if(!facts.length)return '';
  return `<div class="memory-facts-preview"><div class="memory-candidate-form-title">최근 활성 Fact</div>${facts.slice(0,6).map(f=>`<div class="memory-fact-row"><span>${esc(f.scope||'global')}:${esc(f.scope_ref||'default')}</span>${esc(f.statement||'')}</div>`).join('')}</div>`;
}

async function approveMemoryCandidate(id){
  const text=$('candEdit_'+id)?.value||'';
  try{
    await api('/api/memory-candidates/approve',{candidate_id:id, edited_statement:text});
    showToast('후보를 승인해 Fact로 저장했습니다');
    await loadMemoryInbox();
  }catch(e){showToast('승인 실패: '+(e.message||e));}
}

async function rejectMemoryCandidate(id){
  const reason=prompt('거절 사유 (선택)')||'';
  try{
    await api('/api/memory-candidates/reject',{candidate_id:id, reason});
    showToast('후보를 거절했습니다');
    await loadMemoryInbox();
  }catch(e){showToast('거절 실패: '+(e.message||e));}
}

function fillMemoryCandidateExample(){
  const box=$('memoryCandidateJson');
  if(!box)return;
  box.value=JSON.stringify({
    category:'decision', scope:'project', scope_ref:(S.session&&S.session.workspace)||'default',
    statement:'Paperclip 반영은 실행 승인 후에만 진행한다.', source_session_id:(S.session&&S.session.session_id)||null,
    confidence:0.8, recommended_action:'approve', reason:'현재 대화에서 반복 확인된 운영 원칙'
  }, null, 2);
}

async function submitMemoryCandidateJson(){
  const raw=($('memoryCandidateJson')?.value||'').trim();
  if(!raw){showToast('저장할 JSON을 입력하세요');return;}
  let parsed;
  try{parsed=JSON.parse(raw);}catch(e){showToast('JSON 파싱 실패: '+e.message);return;}
  const items=Array.isArray(parsed)?parsed:[parsed];
  try{
    for(const item of items){
      await api('/api/memory-candidates', {
        category:item.category||'knowledge', scope:item.scope||'global', scope_ref:item.scope_ref||'default',
        statement:item.statement||'', source_session_id:item.source_session_id||((S.session&&S.session.session_id)||null),
        source_message_ids:item.source_message_ids||[], confidence:item.confidence, reason:item.reason||'',
        sensitivity:item.sensitivity||'internal', recommended_action:item.recommended_action||'approve', metadata:item.metadata||{}
      });
    }
    showToast(`${items.length}개 후보를 저장했습니다`);
    await loadMemoryInbox();
  }catch(e){showToast('저장 실패: '+(e.message||e));}
}
